src/fwupd_receive_updates.py

```diff
@@ -0,0 +1,281 @@
+        files.
+
+        Keyword argument:
+        updatevm - domain to be checked
+        """
+        cmd = ['qubes-prefs', '--force-root', 'updatevm']
+        p = subprocess.check_output(cmd)
+        self.source = p.decode('ascii').rstrip()
+        if self.source != updatevm:
+            print(
+                'Domain ' + str(self.source) +
+                'not allowed to send dom0 updates',
+                file=sys.stderr
+            )
+            exit(1)
+
+    def _verify_received(self, files_path, regex_pattern):
+        """Checks if sent files are expected.
+
+        Keyword arguments:
+
+        files_path -- absolute path to inspected directory
+        regex_pattern -- pattern of the expected files
+        """
+        # Verify received files
+        for untrusted_f in os.listdir(files_path):
+            if not regex_pattern.match(untrusted_f):
+                raise TypeError(
+                    'Domain ' + self.source + ' sent unexpected file'
+                )
+            f = untrusted_f
+            assert '/' not in f
+            assert '\0' not in f
+            assert '\x1b' not in f
+            path_f = path.join(files_path, f)
+            if os.path.islink(path_f) or not os.path.isfile(path_f):
+                raise TypeError(
+                    'Domain ' + self.source + ' sent not regular file'
+                )
+
+    def _create_dirs(self, *args):
+        """Method creates directories.
+
+        Keyword arguments:
+        *args -- paths to be created
+        """
+        # Create directories for update files
+        qubes_gid = grp.getgrnam('qubes').gr_gid
+        self.old_umask = os.umask(0o002)
+        if args is None:
+            raise ValueError("Creating directories failed, no paths given.")
+        for file_path in args:
+            if not path.exists(file_path):
+                os.mkdir(file_path)
+                os.chown(file_path, -1, qubes_gid)
+                os.chmod(file_path, 0o0775)
+
+    def _extract_archive(self, archive_path, output_path):
+        """Extracts archive file to specified directory.
+
+        Keyword arguments:
+        archive_path -- absolute path to archive file
+        output_path -- absolute output directory
+        """
+        cmd_extract = [
+            "cabextract",
+            "-d",
+            "%s" % output_path,
+            "%s" % archive_path
+        ]
+        shutil.copy(archive_path, FWUPD_DOM0_UPDATES_DIR)
+        p = subprocess.Popen(cmd_extract, stdout=subprocess.PIPE)
+        p.communicate()[0].decode('ascii')
+        if p.returncode != 0:
+            raise Exception(
+                'cabextract: Error while extracting %s.' %
+                archive_path
+            )
+
+    def _gpg_verification(self, file_path):
+        """Inspects GPG signature.
+
+        Keyword argument:
+        file_path -- absolute path to inspected file
+        """
+        cmd_gpg = [
```

Comment:
Packages installed in dom0 are verified with [rpm subprocess](https://github.com/3mdeb/qubes-core-admin-linux/blob/a3098d517f9f85af4ec02db4077ecf2192f340a5/dom0-updates/qubes-receive-updates#L95).

> Generally, it should be preferred to use Pyhton libraries rather than shell commands and parsing outputs.

When we are using the shell commands, we don't have to provide any additional python modules to dom0. This is a security aspect that should be taken into account.

> There must be Python binding for gpg.

There is python module https://pythonhosted.org/python-gnupg/, but it is wrapper for gpg command which uses [gpg subprocess](https://github.com/vsajip/python-gnupg/blob/master/gnupg.py#L934)

Reply:
Let's stick with the Qubes way of doing things and keep it it this way then. The security aspect might be a thing.